content/blog/2019-11-21-bug-finding/index.md

```diff
@@ -0,0 +1,308 @@
+To give a more concrete idea of what kinds of programs we can expected to find bugs in, the authors provide some actual bugs that were found. Generally, these bugs occur when the compiler performs optimizations on programs.
+
+*Wrong Safety Check:*
+```
+(x == c1) || (x < c2)
+```
+is equivalent to
+```
+x < c2
+```
+when `c1` and `c2` are constants and `c1 < c2`. However, Csmith found an example in LLVM where `(x == 0) || (x < -3)` was transformed to `(x < -3)`, even though it's not the case that `0 < -3`. The issue was that LLVM did an unsigned comparison, so the transformation was seemingly safe.
+
+
+*Wrong Analysis:*
+```c++
+1:  static int g[1];
+2:  static int *p = &g[0];
+3:  static int *q = &g[0];
+4:
+5:  int foo (void) {
+6:      g[0] = 1;
+7:      *p = 0;
+8:      *p = *q;
+9:      return g[0];
+10: }
+```
+Here, GCC evaluated this program to 1 instead of 0. The reason is because `q` was accidentally marked as read only, meaning `p` and `q` can't alias (even though they really do). Thus, line 7 looks like dead code because line 8 simply overwrites `*p*`; this is only safe when `p` and `q` don't alias. So line 7 was removed by dead store elimination.
+
+*Wrong Analysis:*
+```c++
+1: void foo(void) {
+2:   int x;
+3:   for (x = 0; x < 5; x++) {
+4:     if (x) continue;
+5:     if (x) break;
+6:   }
+7:   printf("%d", x);
+8: }
+```
+Compiling with LLVM caused this program to print `1` instead of `5`. A loop optimization called "scalar evolution analysis" evaluates loop properties like induction variables and maximum number of iterations. The optimization saw line 5 and incorrectly determined that the loop runs once, meaning `x` must evaluate to `1`.
+
+I find this kind of program interesting because very few programmers in practice would end up writing a function that uses a continue and a break guarded by the same condition. However, when testing optimizations that computes properties of programs, it seems reasonable to test code that never actually runs.
+
+## Design Goals
+The authors provide two design goals for Csmith:
+1. Every randomly generated program must be well formed and have a single interpretation based on the C standard.
+2. Maximize "expressiveness". Expressiveness is the idea that the generated programs should use a wide variety and combinations of language features
+   - For example, Csmith allows for programs with function definitions, global definitions, most C expressions and statements, control flow, arrays, pointers, and more. Some language features that are disallowed include dynamic memory allocation, recursion, and function pointers.
+
+Interestingly, design goal #2 is in direct opposition to design goal #1. The more features we allow to be tested (like arrays and pointers), the easier it is to cause undefined behavior. First, we will discuss at a high level how programs are randomly generated. Then, we will explore how Csmith ensures the generated code doesn't exhibit undefined behavior.
```

Comment:
It might be useful to briefly emphasize _why_ it's important to avoid undefined behavior. (In brief, it's because two different compilers are actually allowed to lead to two different outputs for the same program if it has UB. So Csmith would erroneously flag this as a bug when it's not.)